### PR TITLE
ActiveSync: Support for protocol v14.1 and v14.0 - fixes #212

### LINKED ISSUE #212: ActiveSync: Outlook.com: The user account does not exist in the outlook.com directory
### Reproduction
* Set up an @outlook.com account in Mustang
* Choose ActiveSync as protocol

### Actual result
* Error: "Please login"

image

* Debug output:
```
OAuth2.ts:105 OAuth2ServerError: The user account {EUII Hidden} does not exist in the outlook.com directory. To sign into this application, the account must be added to the directory.
    at _OAuth2.getAccessTokenFromParams (OAuth2.ts:209:13)
    at async _OAuth2.login (OAuth2.ts:103:16)
    at async ActiveSyncAccount.verifyLogin (ActiveSyncAccount.ts:126:9)
    at async checkConfig (checkConfig.ts:8:5)
    at async CheckConfig.svelte:19:3
login @ OAuth2.ts:105
SetupMail.svelte:215 OAuth2LoginNeeded: Please login
    at _OAuth2.login (http://localhost:5454/logic/Auth/OAuth2.ts?t=1726290411389:87:13)
    at async ActiveSyncAccount.verifyLogin (http://localhost:5454/logic/Mail/ActiveSync/ActiveSyncAccount.ts?t=1726290411389:103:9)
    at async checkConfig (http://localhost:5454/logic/Mail/AutoConfig/checkConfig.ts?t=1726290411389:9:5)
    at async http://localhost:5454/frontend/Settings/Setup/Mail/CheckConfig.svelte?t=1726290411389:90:4
```

### Expected result
* Outlook.com OAuth2 login appears in a window
* Can log in
* Account gets set up
* Can fetch emails

### Other results
* It works with Office365 business email accounts like foo@foo.onmicrosoft.com

### PR DESCRIPTION
**Pull request by jermy-c**
- Some accounts use protocol v14.0 e.g. personal accounts
- Missing the `cookie: "DefaultAnchorMailbox={ uri encoded email address }"` causes the HTTP 451 response
- Based on [`neil/activesync-version-14.0`](https://github.com/mustang-im/mustang/tree/neil/activesync-version-14.0)

### CODE REVIEW

**Comment by jermy-c** [comment]
I have gotten `ActiveSync Sync status {number}` errors but I'm guessing it's just because I'm testing and making too many requests. And also I've gotten `Request time out` errors.

**Comment by NeilRashbrook** [comment]
Common Sync error codes:
3. Sync key out of date
4. Request format error
5. Temporary server error
6. invalid request data
7. Change rejected by server
8. Item not found
12. Folder hierarchy changed
16. Retryable server error

**Comment by jermy-c** [comment]
Thank you for that.

These were the codes I managed to record from yesterday.
```
EASError: ActiveSync Sync status 16
    at ActiveSyncAccount.callEAS (ActiveSyncAccount.ts:229:13)
    at async ActiveSyncFolder.makeSyncRequest (ActiveSyncFolder.ts:116:20)
```
	
```
Uncaught (in promise) EASError: ActiveSync Sync status 3
    at ActiveSyncFolder.makeSyncRequest (ActiveSyncFolder.ts:126:13)
```

```
EASError: ActiveSync Sync status 132
    at ActiveSyncAccount.callEAS (ActiveSyncAccount.ts:229:13)
    at async ActiveSyncFolder.makeSyncRequest (ActiveSyncFolder.ts:116:20)
    at async ActiveSyncFolder.queuedSyncRequest (ActiveSyncFolder.ts:101:14)
    at async ActiveSyncFolder.listMessages (ActiveSyncFolder.ts:148:18)
    at async getFirstMessages (saveConfig.ts:90:5)
```

**Comment by NeilRashbrook** [comment]
> EASError: ActiveSync Sync status 16

If we get a lot of these then perhaps we should add an automatic retry after 5 seconds?

> Uncaught (in promise) EASError: ActiveSync Sync status 3

This is weird because all of the callers in the ActiveSync code have an `await`, so is the UI forgetting to `await` something?

> EASError: ActiveSync Sync status 132

This is apparently an internal server error, which used to be returned as an HTTP 500 before protocol 14.0; you're likely to encounter a subsequent status 3.

File: app/logic/Mail/ActiveSync/ActiveSyncAccount.ts [hunk 1]
```diff
@@ -105,9 +114,15 @@ export class ActiveSyncAccount extends MailAccount {
...
     let options: any = {
       throwHttpErrors: false,
-      headers: {},
+      headers: {
+        Cookie: `DefaultAnchorMailbox=${encodeURI(this.emailAddress)}`, // required for v14.0
+      },
     };
-    if (this.oAuth2) {
+    if (this.authMethod == AuthMethod.OAuth2) {
+      let urls = OAuth2URLs.find(a => a.hostnames.includes(this.hostname));
```
Comment (benbucksch) [inline app/logic/Mail/ActiveSync/ActiveSyncAccount.ts:122]:
```suggestion
      let urls = OAuth2URLs.find(a => a.hostnames.includes(this.hostname));
      assert(urls, gt`Could not find OAuth2 config for ${this.hostname}`);
```

File: app/logic/Mail/ActiveSync/ActiveSyncAccount.ts [hunk 2]
```diff
@@ -105,9 +114,15 @@ export class ActiveSyncAccount extends MailAccount {
   async verifyLogin() {
     let options: any = {
       throwHttpErrors: false,
-      headers: {},
+      headers: {
+        Cookie: `DefaultAnchorMailbox=${encodeURI(this.emailAddress)}`, // required for v14.0
+      },
     };
-    if (this.oAuth2) {
+    if (this.authMethod == AuthMethod.OAuth2) {
```
Comment (benbucksch) [inline app/logic/Mail/ActiveSync/ActiveSyncAccount.ts:121]:
```suggestion
    if (this.authMethod == AuthMethod.OAuth2) {
      if (!this.oAuth2) {
```

File: app/logic/Mail/ActiveSync/ActiveSyncAccount.ts [hunk 3]
```diff
@@ -105,9 +114,15 @@ export class ActiveSyncAccount extends MailAccount {
...
+      let urls = OAuth2URLs.find(a => a.hostnames.includes(this.hostname));
+      this.oAuth2 = new OAuth2(this, urls.tokenURL, urls.authURL, urls.authDoneURL, urls.scope, urls.clientID, urls.clientSecret, urls.doPKCE);
+      this.oAuth2.setTokenURLPasswordAuth(urls.tokenURLPasswordAuth);
+      await this.oAuth2.login(true);
```
Comment (benbucksch) [inline app/logic/Mail/ActiveSync/ActiveSyncAccount.ts:125]:
```suggestion
      }
      await this.oAuth2.login(true);
```

(see `if` above)

**Review by benbucksch — CHANGES_REQUESTED** [review]

**Comment by benbucksch** [comment]
> Common Sync error codes: 3. Sync key out of date 4. Request format error 5. Temporary server error 6. invalid request data 7. Change rejected by server 8. Item not found 12. Folder hierarchy changed 16. Retryable server error

Could we please make constants or an enum for that?

File: app/logic/Mail/ActiveSync/ActiveSyncAccount.ts [hunk 4]
```diff
@@ -32,6 +33,7 @@ export class ActiveSyncAccount extends MailAccount {
   listening = false;
   policyKey: Promise<string> | string;
   syncKeyBusy: Promise<any> | null;
+  version: string;
```
Comment (benbucksch) [inline app/logic/Mail/ActiveSync/ActiveSyncAccount.ts:36]: We should call this `protocolVersion`

**Review by benbucksch — APPROVED** [review]

**Review by benbucksch — APPROVED** [review]

**Pull request merged**